cli/src/commands/git/push.rs

```diff
@@ -279,6 +279,22 @@ pub fn cmd_git_push(
             &tx.base_workspace_helper().revset_parse_context(),
         )?)
         .range(&RevsetExpression::commits(new_heads));
+
+    // Only the roots themselves need to be checked if they're being pushed. If
+    // a root is already on the remote, then it is "grandfathered in" and
+    // shouldn't block any descendents from being pushed (unless the descendents
+    // are in private_roots()).
+    let private_commits_revset = revset_util::parse_private_roots_expression(
+        &tx.base_workspace_helper().revset_parse_context(),
+    )?;
+
+    let private_commits = tx
+        .base_workspace_helper()
+        .attach_revset_evaluator(private_commits_revset)?
+        .evaluate_to_commits()?
+        .filter_map(|ch| ch.map(|c| c.id().to_owned()).ok())
+        .collect::<HashSet<_>>();
```

Comment:
Perhaps, you can optimize it by using `containing_fn()`.

Reply:
Oh neat, I didn't know about that. Will change.